### PR TITLE
Enable External GCS.

### PR DESCRIPTION
**Pull request by guoyuhong**
## What do these changes do?
Users may have their own version GCS, e.g. failover supported redis, running outside the head node. In this case, we need to add the option to support external GCS.

In YARN case, when head node is down, we may restart the head node without flushing the  external GCS, so the option to flush external GCS is also added.

## Linter

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/services.py [hunk 1]
```diff
@@ -572,38 +576,54 @@ def start_redis(node_ip_address,
...
-        # Below we use None to indicate no limit on the memory of the
-        # primary Redis shard.
-        redis_max_memory=None,
-        stdout_file=redis_stdout_file,
-        stderr_file=redis_stderr_file)
-    processes.append(p)
-    redis_address = address(node_ip_address, port)
+        redis_address = external_gcs_addresses[0]
+        print("primary_redis_address:", redis_address)
+        [primary_redis_ip, port] = redis_address.split(':')
```
Comment (suquark) [inline python/ray/services.py]: Please use double quotes (lint issue).

**Comment by robertnishihara** [comment]
It should only be necessary to provide the address of the primary shard (which is already done through `--redis-address`), right? Because it can get the shard addresses from the primary shard.

In general, I think we should get rid of the concept of a "head' node and have a separate command for starting the GCS and monitor (which can live on any machine) and a separate `ray start` command for adding new nodes to the cluster.

Example usage could be

```
# Start GCS and monitor on any machine.
ray start-global-services --num-redis-shards=3
```

```
# Add any node to the cluster.
ray start --redis-address=...
```

**Comment by guoyuhong** [comment]
@robertnishihara I agree that we need to get rid of the concept of a "head' node. Maybe some components in head can be treated as actors. 
For current PR, if we only use `--redis-address`, who is responsible to put the key of `NumRedisShards`, `RedirectOutput`, `INCLUDE_JAVA` and `VERSION_INFO` and who is responsible to start the head components?

**Comment by robertnishihara** [comment]
@guoyuhong Right now, the call to `ray start --head ...` creates two things
a) The GCS and monitor
b) A regular Ray node

If we separate these apart, then the Python script that creates the items in a) would set `NumRedisShards` and the other values. Does that answer your question or did I misunderstand it?

I agree it'd be nice if some of the components could be actors.

**Pull request closed without merging**